### PR TITLE
Create controller generator for routing specs

### PR DESCRIPTION
**Pull request by 00dav00**
This PR covers #1160

This creates routing specs for controller generator

/cc @rspec/rspec

### CODE REVIEW

**Comment by samphippen** [comment]
@00dav00 Thanks for this PR! This looks fine to me and I like most of the PR implementation, I do have a few notes. Before we go ahead with this I need to think about whether or not this is a feature we want to support going forward.

I rarely if ever use a routing spec. Do you personally find a lot of value from them?

**Comment by seuros** [comment]
@samphippen I use routing specs for API endpoints since they should follow the specification and not change.

File: lib/generators/rspec/controller/controller_generator.rb [hunk 1]
```diff
@@ -7,8 +7,9 @@ class ControllerGenerator < Base
       argument :actions, :type => :array, :default => [], :banner => "action action"
 
       class_option :template_engine, :desc => "Template engine to generate view files"
-      class_option :controller_specs, :type => :boolean, :default => true
-      class_option :view_specs,       :type => :boolean, :default => true
+      class_option :controller_specs, :type => :boolean, :default => true,  :desc => "Generate controller specs"
+      class_option :view_specs,       :type => :boolean, :default => true,  :desc => "Generate view specs"
+      class_option :routing_specs,    :type => :boolean, :default => true,  :desc => "Generate routing specs"
```
Comment (seuros) [inline lib/generators/rspec/controller/controller_generator.rb]: maybe we could have default : false ?

**Comment by 00dav00** [comment]
@samphippen I agree with @seuros, this specs can be useful when the routes does not follow rest conventions.

**Comment by JonRowe** [comment]
@seuros @00dav00 sure they *can* be useful, but that doesn't mean they should be recommended by default, generated specs have a habit of becoming the default, personally I'd welcome adding a seperate generator, for those that so desire, but not adding them as the default for a controller.

**Comment by seuros** [comment]
@JonRowe what about having the option default to false ?

**Comment by 00dav00** [comment]
@JonRowe I think @seuros 's option may be a good fit, I mean not adding them as default. WDYT @samphippen ?

**Comment by juank-pa** [comment]
I noticed routing specs are being created by the scaffold generator. So isn't this contribution helping us have a sense of completeness? I think this could work but in any case, I'm also open to not having it by default. WDYT? @JonRowe @samphippen

**Comment by samphippen** [comment]
Hi @00dav00 I've found some time to come back to this. I think I like this overall, but I have some questions:

* if I don't specify an action, what happens?
* if I use a scaffold generator to do a full controller backed by a model, what happens?

**Comment by 00dav00** [comment]
Hi @samphippen, thanks for puting some time into this, highly appreciated.

1) `if I don't specify an action, what happens?`

If `--routing-specs` is passed to the generator but no action is given the spec file will be created but will contain no specs. Should I change this behavior to create route specs for the CRUD operations (like the scaffold routing generator) or the creation of this file should be skipped (like in the method for view specs)?

2) `if I use a scaffold generator to do a full controller backed by a model, what happens?`

The scaffold generator is creating the routing specs by default. In opposition with the current PR, the scaffold generator will only skip the routing specs if the flag `--no-routing-specs` is provided. Also if no actions are sent the CRUD operations will be used to create the routing specs. I think this should be consistent between the two generators, WDYT @samphippen? Which approach should be used?

**Comment by 00dav00** [comment]
@samphippen This has been here for a while now, do you think we will be able to move it forward or you consider it is not required?

File: lib/generators/rspec/controller/controller_generator.rb [hunk 2]
```diff
@@ -7,8 +7,9 @@ class ControllerGenerator < Base
       argument :actions, :type => :array, :default => [], :banner => "action action"
 
       class_option :template_engine, :desc => "Template engine to generate view files"
-      class_option :controller_specs, :type => :boolean, :default => true
-      class_option :view_specs,       :type => :boolean, :default => true
+      class_option :controller_specs, :type => :boolean, :default => true,  :desc => "Generate controller specs"
+      class_option :view_specs,       :type => :boolean, :default => true,  :desc => "Generate view specs"
+      class_option :routing_specs,    :type => :boolean, :default => false,  :desc => "Generate routing specs"
```
Comment (JonRowe) [inline lib/generators/rspec/controller/controller_generator.rb]: Double spacing here, unless you were intending to align the `:desc`, in which case the latst needs a space removed.

File: lib/generators/rspec/controller/controller_generator.rb [hunk 3]
```diff
@@ -29,6 +30,13 @@ def generate_view_specs
                    File.join("spec", "views", file_path, "#{@action}.html.#{options[:template_engine]}_spec.rb")
         end
       end
+
+      def generate_routing_spec
+        return unless options[:routing_specs]
+
```
Comment (JonRowe) [inline lib/generators/rspec/controller/controller_generator.rb]: This should probably return is actions is empty like views...

File: lib/generators/rspec/controller/templates/routing_spec.rb [hunk 4]
```diff
@@ -0,0 +1,13 @@
+require 'rails_helper'
+
+<% module_namespacing do -%>
+RSpec.describe <%= class_name %>Controller, <%= type_metatag(:routing) %> do
```
Comment (JonRowe) [inline lib/generators/rspec/controller/templates/routing_spec.rb]: This describe should be a doc string, we don't need the constant here.

**Comment by 00dav00** [comment]
Closing this one in favor of #2134

**Pull request closed without merging**